function-sidecar/pkg/dispatcher/grpc/grpc.go

```diff
@@ -114,14 +177,23 @@ func NewGrpcDispatcher(port int, timeout time.Duration) (dispatcher.Dispatcher,
 		return nil, err
 	}
 
-	fnStream, err := function.NewMessageFunctionClient(conn).Call(context.Background())
-	if err != nil {
-		return nil, err
+	result := &grpcDispatcher{
+		client:         function.NewMessageFunctionClient(conn),
+		input:          make(chan message.Message, 100),
+		output:         make(chan message.Message, 100),
+		closed:         make(chan struct{}),
+		streams:        make(map[interface{}]ongoingStream),
+		windowingFactory: unmarshallFactory(),
+		correlation: func(m message.Message) interface{} {
+			// Correlate by correlationId header by default if present
+			if c, ok := m.Headers()[correlationId] ; ok {
+				return c[0]
+			} else {
+				return nil
```

Comment:
I wonder if we should correlate by partition too to preserve kafka ordering guarantees.

Reply:
Not sure what you mean here. Broker partitioning has already happened upstream, so this partitioning is finer grained.
Imagine we have 2 kafka partitions and one pod gets even orderIds, while the other gets odd ones.

Then _this_ partitioning can be by, say, orderId itself. But this particular pod may only see odd orderIds anyway

Reply:
I don't understand how broker partitioning can already have happened. The sidecar consumer gets a message and sends it to the gRPC dispatcher. But if multiple partitions are associated with the consumer, the message could be from any partition.

Reply:
yes, I agree we would ideally de-multiplex before invoking the RPC calls so that for any single RPC call, the messages are from a single partition, but perhaps that's something to address after this PR, since that same situation applies today.... at least we'd be making progress on the windowing as opposed to always sending unbounded streams

Reply:
Ok, fair enough. Thanks.